File: apps/server/routes/telemetry.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import type { Hono } from "hono";

import { TtlCache, type ICache } from "@quip/core/cache";
import type { DatabaseAdapter } from "@quip/core/db/adapter";
import { parseDispatchAttemptsApiResponse } from "@quip/core/miner-api";
import { resolveSelfMinerRestUrl } from "@quip/core/resolve-miner-rest";
import type {
  CurrentDispatch,
  MiningAttempt,
  NodeDescriptorRecord,
  NodeInfo,
  NodesSnapshot,
  TelemetryResponse,
  ValidatorAuthorshipRecord,
} from "@quip/shared/telemetry";
import { getGeoIpEnricher } from "../geo-ip";

// Cap on submission rows surfaced in /api/telemetry. The MyNode panel
// shows the most recent and offers click-through to the modal for older
// ones, so a deeper history is just bytes on the wire. 20 fits one
// screenful of rows comfortably.
const RECENT_MINING_SUBMISSIONS_LIMIT = 20;
// "Online" threshold for the Active Validators table. A validator counts
// as online when its most recent authored head is within this window of
// the request wall-clock. 3 minutes is roughly 30x the 6s block time on
// quip-protocol-rs spec 101 — short enough to catch operator outages,
// long enough that BABE slot skips don't briefly flap a healthy node.
const VALIDATOR_ONLINE_WINDOW_MS = 3 * 60 * 1000;
// The telemetry payload is global to the deployment (no per-viewer state), so
// one process-wide snapshot serves every concurrent poller. A short TTL keeps
// staleness well under the indexer's own write cadence while collapsing N
// viewers' per-poll recompute into one. The server-side cache is independent
// of the `no-store` header below, which only stops *browsers* from caching.
const DEFAULT_CACHE_TTL_MS = 1000;

const SNAPSHOT_KEY = "telemetry";

interface TelemetryDeps {
  db: DatabaseAdapter;
  validatorRpcUrls: string[];
  // Injected clock — drives the snapshot timestamps (and the default cache's
  // freshness check), so tests can pin time deterministically.
  now?: () => number;
  // 0 disables the cache (every request rebuilds); used by tests for an
  // uncached baseline. Ignored when an explicit `cache` is injected.
  cacheTtlMs?: number;
  // The snapshot cache. Defaults to an in-process TtlCache; injectable so tests
  // (or alternative deployments) can supply their own ICache implementation.
  cache?: ICache<TelemetryResponse>;
}

export function registerTelemetryRoute(app: Hono, deps: TelemetryDeps): void {
  const { db, validatorRpcUrls } = deps;
  const now = deps.now ?? Date.now;
  const cacheTtlMs = deps.cacheTtlMs ?? DEFAULT_CACHE_TTL_MS;
  const cache = deps.cache ?? new TtlCache<TelemetryResponse>({ ttlMs: cacheTtlMs, now });

  app.get("/api/telemetry", async (c) => {
    const snapshot = await cache.read(SNAPSHOT_KEY, buildSnapshot);
    // Telemetry is a moving target — every poll returns different counters.
    // Without `no-store`, browsers can apply heuristic freshness to an opaque
    // JSON body and serve a cached response after a refocus, masking real
    // updates from the indexer.
    c.header("Cache-Control", "no-store");
    return c.json(snapshot);
  });

  async function buildSnapshot(): Promise<TelemetryResponse> {
    const [
      blocks,
      selfAddress,
      indexer,
      chainHead,
      babeEpoch,
      babeAuthorities,
      chainMiners,
      recentDifficulty,
      allHardware,
      authorship,
      nodeDescriptors,
    ] = await Promise.all([
      // Page-1 default; the UI can request later pages once pagination lands.
      db.getRecentBlocks(500, 0),
      db.getSelfAddress(),
      db.getIndexerObservability(),
      db.getChainHead(),
      db.getCurrentBabeEpoch(),
      db.getActiveBabeAuthorities(),
      db.getChainMiners(),
      db.getRecentDifficulty(50),
      db.getAllMinerHardware(),
      db.getValidatorAuthorship(),
      db.getAllNodeDescriptors(),
    ]);

    // Recent submissions by the locally-polled miner — drives the
    // "Recent Performance" panel. Empty until either selfAddress
    // resolves (first /status poll) or the miner emits its first
    // submission. The indexer keys by selfAddress so multi-miner
    // dashboards never see other miners' submissions here.
    const recentMiningSubmissions = selfAddress
      ? await db.getRecentMiningSubmissions(selfAddress, RECENT_MINING_SUBMISSIONS_LIMIT)
      : [];

    // Lifetime "Problems Attempted" — distinct solution_ids with iterations
    // recorded. Counted at the DB so the value isn't bounded by the recent-
    // submissions window above.
    const selfProblemsAttempted = selfAddress
      ? await db.countMiningSubmissionsWithAttempts(selfAddress)
      : 0;

    // Hardware lookup feeds two downstream concerns: the chain-miner
    // join (every row) and the current-dispatch fetch (self only). Build
    // once.
    const hardwareByAccount = new Map(allHardware.map((h) => [h.accountId, h]));

    // Current-dispatch resolution. The in-flight problem is the global
    // solution_number = `LatestQBlockId + 1`, surfaced through
    // `chain_head.winning_solutions_count` for API compatibility. The miner
    // grinds that solution_number; the prior one is the just-completed
    // problem. Skip when the substrate worker hasn't written the count yet.
    const minerHardware = selfAddress ? hardwareByAccount.get(selfAddress) : undefined;
    const minerInternalId = minerHardware?.miners[0]?.id ?? null;
    const winningSolutionsCount = chainHead?.winningSolutionsCount ?? null;
    const currentDispatch: CurrentDispatch | null =
      selfAddress && minerInternalId && winningSolutionsCount !== null
        ? await resolveCurrentDispatch(minerInternalId, winningSolutionsCount + 1)
        : null;

    // Project per-account registry descriptors into the legacy
    // NodesSnapshot shape so the Compute Available view's TFLOPS/PFLOPS
    // surfaces keep their existing consumer contract. The descriptor
    // pipeline replaces the v0.2 HTTP-fanout survey, but the snapshot
    // shape is unchanged.
    const projected = projectDescriptorsToSnapshot(nodeDescriptors);
    // Geo-IP enrich the projection: resolve each node's `publicHost` to a
    // lat/lng via DNS + MaxMind. Cache-amortised, so warm calls are
    // synchronous in practice. No-op when geo backends are unavailable
    // (the map then renders no markers).
    const enricher = await getGeoIpEnricher();
    const nodes = await enricher.enrichSnapshot(projected);

    // Join chain_miners → miner_hardware on accountId so the UI can render
    // a per-row "telemetry node" link without a second fetch. Today only
    // self has a miner_hardware row (source='self'); future peer-query and
    // chain-surface upgrades populate other entries.
    const enrichedMiners = chainMiners.map((m) => ({
      ...m,
      telemetryNodeAddress: hardwareByAccount.get(m.accountId)?.nodeId ?? null,
      hardware: hardwareByAccount.get(m.accountId) ?? null,
    }));

    // Join the active BABE authority set → per-validator authorship stats.
    // Validators that haven't authored a head the indexer has seen surface
    // with 0 counters and `online: false`; the row still appears in the
    // table so operators see their full authority set, not just the busy
    // ones.
    const authorshipByAccount = new Map(authorship.map((a) => [a.accountId, a]));
    const nowMs = now();
    const validators: ValidatorAuthorshipRecord[] = babeAuthorities.map((a) => {
      const stats = authorshipByAccount.get(a.accountId);
      const lastAuthoredAt = stats?.lastAuthoredAt ?? null;
      const ageMs = lastAuthoredAt ? nowMs - Date.parse(lastAuthoredAt) : Infinity;
      return {
        accountId: a.accountId,
        blocksAuthored: stats?.blocksAuthored ?? 0,
        blocksAuthoredWithPow: stats?.blocksAuthoredWithPow ?? 0,
        lastAuthoredBlock: stats?.lastAuthoredBlock ?? null,
        lastAuthoredAt,
        online: ageMs < VALIDATOR_ONLINE_WINDOW_MS,
      };
    });

    return {
      blocks,
      selfAddress,
      indexer,
      serverTime: new Date(now()).toISOString(),
      chainHead,
      babeEpoch,
      babeAuthorities,
      chainMiners: enrichedMiners,
      recentDifficulty,
      validators,
      nodes,
      nodeDescriptors,
      recentMiningSubmissions,
      selfProblemsAttempted,
      currentDispatch,
    } satisfies TelemetryResponse;
  }

  /**
   * Resolve which global solution_number the panel shows + whether it's
   * in-flight (MR !105).
   *
   * The miner is grinding `currentSolutionNumber` (= Σ proofsWon + 1); its
   * prior problem `currentSolutionNumber - 1` (= Σ proofsWon) was won
   * network-wide and is finished. Probe both: if the in-flight directory
   * has iterations the miner is actively working it; otherwise fall back to
   * the just-completed one for the brief window before the next directory
   * is written.
   *
   * Best-effort throughout: any failure path that ends with no iterations
   * returns null and the panel renders the empty state (which explains
   * "between dispatches" / "miner unreachable").
   */
  async function resolveCurrentDispatch(
    minerId: string,
    currentSolutionNumber: number,
  ): Promise<CurrentDispatch | null> {
    const [curAttempts, prevAttempts] = await Promise.all([
      fetchSolutionAttempts(minerId, currentSolutionNumber),
      currentSolutionNumber > 1
        ? fetchSolutionAttempts(minerId, currentSolutionNumber - 1)
        : Promise.resolve<MiningAttempt[]>([]),
    ]);
    if (curAttempts.length > 0) {
      return {
        solutionNumber: currentSolutionNumber,
        attempts: curAttempts,
        status: "in-flight",
      };
    }
    if (prevAttempts.length > 0) {
      return {
        solutionNumber: currentSolutionNumber - 1,
        attempts: prevAttempts,
        status: "completed",
      };
    }
    return null;
  }

  async function fetchSolutionAttempts(
    minerId: string,
    solutionNumber: number,
  ): Promise<MiningAttempt[]> {
    const baseUrl = resolveSelfMinerRestUrl(validatorRpcUrls);
    if (!baseUrl) return [];
    const params = new URLSearchParams({
      miner_id: minerId,
      solution_number: String(solutionNumber),
    });
    const url = `${baseUrl}/api/v1/mining/attempts?${params.toString()}`;
    try {
      const res = await fetch(url, {
        headers: { accept: "application/json" },
        signal: AbortSignal.timeout(4000),
      });
      if (!res.ok) return [];
      const parsed = (await res.json()) as { success?: boolean; data?: unknown };
      if (parsed && typeof parsed === "object" && parsed.success === false) return [];
      return parseDispatchAttemptsApiResponse(parsed?.data ?? parsed);
    } catch {
      return [];
    }
  }

  // Project the descriptor-worker's per-account rows into the legacy
  // NodesSnapshot shape the Compute Available view already consumes.
  // Field ownership per DASHBOARDPLAN.md:
  //   - chain-derived: address (SS58), firstSeen, lastSeen (block timestamps)
  //   - operator self-asserted: nodeName, publicHost, runtime, miners,
  //     systemInfo (from descriptor)
  //   - dashboard placeholders: status="active" (we have no liveness signal
  //     yet — the chain doesn't beat), lastHeartbeat=null (same)
  function projectDescriptorsToSnapshot(records: NodeDescriptorRecord[]): NodesSnapshot | null {
    if (records.length === 0) return null;
    const nodes: Record<string, NodeInfo> = {};
    let mostRecentObservedAt = "";
    for (const r of records) {
      const d = r.descriptor;
      nodes[r.accountId] = {
        address: r.accountId,
        // No chain-side liveness signal; treat any registered descriptor
        // as "active" until we have a heartbeat-equivalent surface.
        status: "active",
        firstSeen: r.firstBlockTimestamp,
        lastSeen: r.blockTimestamp,
        lastHeartbeat: null,
        nodeName: d.nodeName,
        publicHost: d.publicHost,
        publicPort: d.publicPort,
        autoMine: d.autoMine,
        logLevel: d.logLevel,
        runtime: d.runtime,
        miners: d.miners,
        systemInfo: d.systemInfo,
      };
      if (r.observedAt > mostRecentObservedAt) mostRecentObservedAt = r.observedAt;
    }
    return {
      updatedAt: mostRecentObservedAt || new Date().toISOString(),
      nodeCount: records.length,
      // No liveness distinction yet; will refine once we have a chain-side
      // heartbeat or join against recent validator authorship.
      activeCount: records.length,
      nodes,
    };
  }
}

File: crates/quip-dashboard/src/http/telemetry.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
use super::{HttpState, routes::ApiError};
use axum::{
    body::{Body, Bytes},
    extract::State,
    http::header,
    response::{IntoResponse, Response},
};
use dashboard_model::{TelemetryFiles, TelemetryResponse, ValidatorAuthorshipRecord};
use std::{collections::BTreeMap, time::Duration};
use tokio::time::Instant;
#[derive(Default)]
pub(super) struct SnapshotCache {
    payload: Option<(Instant, Bytes)>,
}
pub(super) async fn get(State(state): State<HttpState>) -> Result<Response, ApiError> {
    // The mutex is also the single-flight gate. Bytes keep their quota until
    // both the cache and every in-flight HTTP body release their shared owner.
    let mut cache = state.cache.lock().await;
    let bytes = if let Some((expires, bytes)) = &cache.payload {
        if *expires > Instant::now() {
            Some(bytes.clone())
        } else {
            None
        }
    } else {
        None
    };
    let bytes = if let Some(bytes) = bytes {
        bytes
    } else {
        cache.payload = None;
        let snapshot = Box::pin(build(&state)).await?;
        let bytes = super::response::encode(&state, &snapshot)?;
        cache.payload = Some((Instant::now() + Duration::from_secs(1), bytes.clone()));
        bytes
    };
    Ok((
        [
            (header::CONTENT_TYPE, "application/json"),
            (header::CACHE_CONTROL, "no-store"),
        ],
        Body::from(bytes),
    )
        .into_response())
}
async fn build(state: &HttpState) -> Result<TelemetryResponse, ApiError> {
    let db = &state.store;
    let (
        self_address,
        indexer,
        chain_head,
        babe_epoch,
        babe_authorities,
        mut chain_miners,
        mut recent_difficulty,
        hardware,
        authorship,
        mineable_topologies,
    ) = tokio::try_join!(
        db.get_self_address(),
        db.get_indexer_observability(),
        db.get_chain_head(),
        db.get_current_babe_epoch(),
        db.get_active_babe_authorities(),
        db.get_chain_miners(),
        db.get_recent_difficulty(50),
        db.get_all_miner_hardware(),
        db.get_validator_authorship(),
        db.get_mineable_topologies(),
    )?;
    let self_address = self_address.or_else(|| state.operator_account.clone());
    let (recent_mining_submissions, self_problems_attempted) = if let Some(account) = &self_address
    {
        tokio::try_join!(
            db.get_recent_mining_submissions(account, 20),
            db.count_mining_submissions_with_attempts(account)
        )?
    } else {
        (Vec::new(), 0)
    };
    let hardware: BTreeMap<_, _> = hardware
        .into_iter()
        .map(|row| (row.account_id.clone(), row))
        .collect();
    for miner in &mut chain_miners {
        miner.hardware = hardware.get(&miner.account_id).cloned();
        miner.telemetry_node_address = miner.hardware.as_ref().map(|row| row.node_id.clone());
    }
    let now = (state.clock)();
    let validators = babe_authorities
        .iter()
        .map(|authority| {
            let stats = authorship
                .iter()
                .find(|row| row.account_id == authority.account_id);
            let last_authored_at = stats.and_then(|row| row.last_authored_at.clone());
            let online = last_authored_at
                .as_deref()
                .and_then(|value| chrono::DateTime::parse_from_rfc3339(value).ok())
                .is_some_and(|time| now.signed_duration_since(time).num_milliseconds() < 180_000);
            ValidatorAuthorshipRecord {
                account_id: authority.account_id.clone(),
                blocks_authored: stats.map_or(0, |row| row.blocks_authored),
                blocks_authored_with_pow: stats.map_or(0, |row| row.blocks_authored_with_pow),
                last_authored_block: stats.and_then(|row| row.last_authored_block.clone()),
                last_authored_at,
                online,
            }
        })
        .collect();
    if let Some(topology) = mineable_topologies
        .iter()
        .find(|row| row.is_default)
        .map(|row| &row.topology_hash)
    {
        recent_difficulty.retain(|row| row.topology_hash.as_ref() == Some(topology));
    }
    let miner_current_dispatch = miner_dispatch_url(self_address.as_deref());
    Ok(TelemetryResponse {
        self_address,
        indexer,
        server_time: state.now(),
        chain_head,
        babe_epoch,
        babe_authorities,
        chain_miners,
        recent_difficulty,
        mineable_topologies,
        validators,
        recent_mining_submissions,
        self_problems_attempted,
        files: TelemetryFiles {
            qblocks_manifest: "/files/qblocks/metadata.json".to_owned(),
            nodes_snapshot: "/files/nodes/snapshot.json".to_owned(),
            miner_current_dispatch,
        },
        capabilities: dashboard_model::Capabilities {
            miner_dispatch: state.miner_dispatch,
        },
    })
}

/// Static URL of `account`'s dispatch document, or `None` without an account.
fn miner_dispatch_url(account: Option<&str>) -> Option<String> {
    account.map(|account| format!("/files/miners/{account}/current-dispatch.json"))
}

#[cfg(test)]
mod tests {
    use super::miner_dispatch_url;

    /// The pointer names the account's file, and is absent without an account.
    #[test]
    fn the_dispatch_pointer_follows_the_miner_file_layout() {
        assert_eq!(
            miner_dispatch_url(Some("5G8Ack")).as_deref(),
            Some("/files/miners/5G8Ack/current-dispatch.json")
        );
        assert_eq!(miner_dispatch_url(None), None);
    }
}
